design: add sub-states

dtantsur opened the pull request:
This design proposes a new `BareMetalHost` field `SubState` to specify more
precisely at which point in a BMH lifecycle it is.

dtantsur commented:
/assign @dhellmann

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+
+  * IF `Node.deploy_step` is null OR `Node.deploy_step["step"]` is "deploy":
+
+    SubState is `Preparing Deployment`.
```
```suggestion
    SubState is `Preparing to write image`.
```

dtantsur replied:
Works for me, just keep in mind that for non-fast-track case it includes booting the agent.

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+
+  * ELSE:
+
+    SubState is `Configuring`.
```
What does "configuring" mean? What is being configured?

dtantsur replied:
Bootloader (if needed), BMC for local boot and a few other small tasks.

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+    SubState is `Configuring`.
+
+<!-- dtantsur: it's possible to be more granular, but it may not be justified,
+especially since only the write_image step runs noticable amount of time -->
```
Yeah, depending on how long some of the transitions take it's entirely possible we would skip over some states if the polling is out of sync with the state change within Ironic. So larger steps are better, but this extra detail is good.

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+
+  * IF `Node.clean_step` is null:
+
+    SubState is `Preparing Configuration`.
```
Is this about preparing to configure the RAID controller, then?

dtantsur replied:
It's the same as "Preparing cleaning" above, but I'm trying to avoid using "cleaning" to mean anything except for "cleaning disks". This includes booting the agent (if without fast-track) and doing other small preparation items.

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+
+<!-- dtantsur: we could split further between deleting the previous RAID
+configuration and creating a new one. I'm not sure it brings enough benefit,
+to be honest -->
```
Probably not.

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+
+### Risks and Mitigations
+
+None?
```
Saying that the new value is not part of the API might impose a risk if someone doesn't realize that and does in fact try to build a client that expects specific values in that field.

dtantsur replied:
Is there a potential mitigation for that?

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+
+### Work Items
+
+<!-- dtantsur: also unsure here -->
```
We'll need to update the provisioner code in https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go to set the right activity value from methods where changes in the activity value are likely to be useful.

If we decide to make the values part of the API, then we need to define an enum and update the CRD to include the values.

dtantsur replied:
I guess an enum is better for API consumeability, but is it fine to add/remove states later? (Removing is less important as we can have states that are not used).

dhellmann replied:
Yes, it's OK to extend the enum over time.

dtantsur commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+
+<!-- dtantsur: I'm terribly unfamiliar with the code, help appreciated -->
+
+* New field `SubState` is added to `BareMetalHost`
```
`CurrentActivity` sounds great, will change.

dtantsur commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+
+  * IF `Node.clean_step` is null:
+
+    SubState is `Preparing Cleaning`.
```
This is the stage where we prepare ramdisk boot (if it's not already booted), wait for this boot and figure out clean steps to run. "Reset storage configuration" sounds too scare to me. Would it be so bad to introduce "cleaning" to the API in the narrow sense of "Cleaning hard drives"?

dhellmann replied:
Talking about "cleaning" before provisioning seems odd. I guess we'll end up in the same state when deprovisioning the host? "Preparing to update host configuration" would handle both cases.

dtantsur commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+  * IF `Node.clean_step["step"]` in "erase\_devices",
+    "erase\_devices\_metadata":
+
+    SubState is `Cleaning Disks`.
```
This will be misleading since by default we don't wipe the information, only partitioning table. "Deleting disk metadata"?

dhellmann replied:
Yes, changing "content" to metadata works.

dtantsur commented:
/cc @zaneb

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+  * IF `Node.clean_step["step"]` in "erase\_devices",
+    "erase\_devices\_metadata":
+
+    Activity is `Cleaning disks`.
```
```suggestion
    Activity is `Deleting disk metadata`.
```

dtantsur replied:
In theory, we can also enable full cleaning (this is what `erase_devices` is about). But okay, we can change this for now.

dhellmann reviewed (commented):
This is looking good.

Is it possible for us to predict in advance (before calling Ironic) how many high-level steps would be involved in provisioning or deprovisioning? We don't need every ironic state transition, but if we knew how many of these "activity" steps there were going to be then we could provide some sort of progress information that a UI could use. Obviously the steps might take different lengths of time in different cases, but knowing that you're on step 4/6 is better than nothing.

andfasano commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+an Ironic Node object. The rules are defined in pseudo-code.
+
+This proposal introduces a concept of "cleaning" with a narrower definition
+that in Ironic. While in Ironic "cleaning" refers to any pre-deployment
```
nit: than

andfasano commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,155 @@
+    "switch\_to\_tenant\_network", "boot\_instance":
+
+    Activity is `Rebooting into operating system`.
+
```
Probably a default case could be useful to empty `CurrentActivity` when none of the above conditions match, otherwise the field could report an incorrect situation?

dtantsur replied:
It's implicit in my proposal, I'll make it explicit.

dtantsur commented:
> Is it possible for us to predict in advance (before calling Ironic) how many high-level steps would be involved in provisioning or deprovisioning? We don't need every ironic state transition, but if we knew how many of these "activity" steps there were going to be then we could provide some sort of progress information that a UI could use. Obviously the steps might take different lengths of time in different cases, but knowing that you're on step 4/6 is better than nothing.

This document pretty much defines what you're asking for, I think.

dtantsur commented:
Okay, I've added a way to estimate progress. It's based on calculations for cleaning and on prior knowledge for deployment.

Also added some information about Ironic for the readers who are not familiar with it.

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,184 @@
+* The `ProvisionStatus` structure is extended with a new string field
+  `CurrentActivity` and a new integer field `Progress`.
+* The `Progress` value is a number from 0 to 100 that corresponds to estimated
+  procentage of the work that has been done so far.
```
```suggestion
  percentage of the work that has been done so far.
```

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,184 @@
+
+### Non-Goals
+
+* Provide an access to the Ironic Node `provision_state` field.
```
```suggestion
* Provide access to the Ironic Node `provision_state` field.
```

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,184 @@
+from within it. When *fast track* mode is on, the ramdisk is booted only once
+and inspection, cleaning and deployment happen in a sequence.
+
+Cleaning and deployment processed are split into potentially pluggable *clean*
```
```suggestion
Cleaning and deployment processes are split into potentially pluggable *clean*
```

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,184 @@
+* The values of the fields are re-calculated based on the result described
+  below on each reconciliation.
+
+### Ironic steps basics
```
This background section was very helpful, thank you for adding it.

dhellmann commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,184 @@
+
+  * IF `Node.deploy_step` is null OR `Node.deploy_step["step"]` is "deploy":
+
+    Activity is `Preparing to write image`. Progress is 30.
```
These progress values would apply after each step is completed, right? So when we move from preparing-to-write to writing, we would set the progress to 30?

dtantsur replied:
No, I'm trying to write an algorithm here, so when this condition is true, progress is set to 30.

dhellmann replied:
OK, that's a little confusing. What is happening before we get to preparing-to-write that accounts for 30 percent of the total work? Booting the agent?

dtantsur replied:
The whole automated cleaning (since for us it's a part of the deployment).

dhellmann reviewed (commented):
Nice, you took the progress monitoring farther than I expected. One question and a couple of typo fixes inline.

stbenjam commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,185 @@
+
+Ironic has three major processes that are of interest for us:
+
+* *Inspection* is a process of populating hardware information.
```
Are there any relevant substates for inspection? Should inspection be figured into the percentage complete?

dhellmann replied:
From metal3's perspective, inspection is a separate set of steps that come before provisioning can proceed. So, we could show % complete of inspection, then reset that for provisioning.

dtantsur replied:
Unfortunately, there are no steps for inspection.

dhellmann replied:
OK, maybe we can do something about that in the future but for now we can ignore it.

andybraren commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,185 @@
+
+* The `ProvisionStatus` structure is extended with a new string field
+  `CurrentActivity` and a new integer field `Progress`.
+* The `Progress` value is a number from 0 to 100 that corresponds to estimated
```
Do we think the time at which the step began/finished could be helpful for debugging? See design concept 2 of what this might look like in a GUI.

dtantsur replied:
I think it's potentially useful, although I have concerns about concept 2 (see my other message). However, I'd prefer to limit the scope of work for this design, because there's truly no limit to improving it :) Similarly, I'd start with an easier UI concept.

andybraren reviewed (commented):
+1 to this proposal's motivation. Providing users with more insight into the long-running `Provisioning` process would be a great UX improvement.

I started brainstorming how these new sub-states might be reflected in a GUI. Concept 1 below is what this proposal as-written might look like, with the overall progress and each sub-state shown. Much better than what we have today. 👍

For the sake of discussion, I also mocked up a second concept that lists every sub-state and the time at which each step finished. I've seen some products use this step-based approach to avoid some of the pitfalls of trying to approximate percentages. 9 steps seems a little overwhelming to my eye (maybe some are quick/combinable?) but maybe this approach has debuggability benefits if something goes wrong? 🤷‍♂️

![metal3-138-provisioning-statuses](https://user-images.githubusercontent.com/9122899/94493279-346c7500-01ba-11eb-8780-04c5b750351f.png)

FYI @nirfarkas

dtantsur commented:
>  I also mocked up a second concept that lists every sub-state and the time at which each step finished

I would recommend avoiding making any assumption about the exact states that will be executed and their order. For example, RAID and firmware settings are optional. I'm must more in favour of the concept 1 for this reason (great job btw!)

dtantsur commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,185 @@
+
+  * IF `Node.deploy_step["step"]` is "prepare\_instance\_boot":
+
+    Activity is `Configuring boot`. Progress is 75.
```
It's about configuring the bootloader (in our case calling `efibootmgr` from within the machine) and telling the BMC to boot from disk. "Preparing to reboot" is closer to the next step.

andybraren commented:
@dtantsur Makes sense, sounds good to me!

honza commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,185 @@
+
+    Activity is `Rebooting into operating system`. Progress is 90.
+
+* ELSE Activity is an empty string and Progress it not changed.
```
Typo 

```suggestion
* ELSE Activity is an empty string and Progress is not changed.
```

zaneb commented on design/baremetal-operator/sub-states.md:
```diff
@@ -0,0 +1,185 @@
+
+    Activity is `Configuring RAID`.
+
+  * IF `Node.clean_step["interface"]` == "bios":
```
One issue we have with the patch to add RAID support is that the Node state returns to manageable after cleaning. To ensure we clean exactly once, the patch adds a new "Preparing" state (in which we kick off cleaning and immediately switch to the "Provisioning" state). It would be nice to either not have to add this new state, or to put all of the cleaning steps into the Preparing state - as it stands it's pretty misleading to the user. I don't think any changes to this proposal are required to do this, but it might be worth mentioning as one of the goals.

demonCoder95 commented:
/remove-lifecycle-stale
What's holding this up? This looks like a really  nice addition to the operator and should be brought up more often.

dtantsur commented:
/remove-lifecycle stale

fmuyassarov commented:
If there are no feedback let;s have a week of lazy consensus.

fmuyassarov commented:
@dhellmann , @zaneb , @andfasano can we get this merged this if it looks okay? Because, we have agreed in the last community meeting that we would merge the proposal if there no more comments during the next one week. And @dtantsur  brought up the proposal today as well, and seems there is no further comments/objections to be addressed.
If you think it is good already, it would be nice to get it merged and move on with implementation. Thanks

andfasano commented:
Sorry for the late feedback, I'm fine with the proposal.

dhellmann commented:
> @dhellmann , @zaneb , @andfasano can we get this merged this if it looks okay? Because, we have agreed in the last community meeting that we would merge the proposal if there no more comments during the next one week. And @dtantsur brought up the proposal today as well, and seems there is no further comments/objections to be addressed.
> If you think it is good already, it would be nice to get it merged and move on with implementation. Thanks

It looks fine to me. @fmuyassarov, was the decision that it's ready to merge today or that we should start the lazy consensus period?

/approve

fmuyassarov commented:
> > @dhellmann , @zaneb , @andfasano can we get this merged this if it looks okay? Because, we have agreed in the last community meeting that we would merge the proposal if there no more comments during the next one week. And @dtantsur brought up the proposal today as well, and seems there is no further comments/objections to be addressed.
> > If you think it is good already, it would be nice to get it merged and move on with implementation. Thanks
> 
> It looks fine to me. @fmuyassarov, was the decision that it's ready to merge today or that we should start the lazy consensus period?
> 
> /approve

It was agreed last week that we put lazy consensus of a week, which means it ends today. So, I think it is good to go in.

dhellmann commented:
Let's get this moving, then.

/lgtm
